File: app/src/main/java/com/jameschamberlain/footballteamtracker/FileUtils.java
package com.jameschamberlain.footballteamtracker;

import android.content.Context;
import android.util.Log;

import com.jameschamberlain.footballteamtracker.fixtures.Fixture;
import com.jameschamberlain.footballteamtracker.fixtures.Score;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.time.LocalDateTime;
import java.util.ArrayList;

public class FileUtils {

    private static final String TAG = "FileUtils";

    private static final String FIXTURES_FILE_NAME = "fixtures.txt";
    private static final String TEAM_FILE_NAME = "team.txt";
    private static final String PLAYERS_FILE_NAME = "players.txt";
    private static File fixturesFile;
    private static File teamFile;
    private static File playersFile;
    private static String fixturesPath;
    private static String teamPath;
    private static String playersPath;


    public static void checkFiles(Context context) {
        fixturesPath = context.getFilesDir() + FIXTURES_FILE_NAME;
        fixturesFile = new File(fixturesPath);
        if (!fixturesFile.exists()) {
            findFile(fixturesFile, fixturesPath);
        }

        teamPath = context.getFilesDir() + TEAM_FILE_NAME;
        teamFile = new File(teamPath);
        if (!teamFile.exists()) {
            findFile(teamFile, teamPath);
        }

        playersPath = context.getFilesDir() + PLAYERS_FILE_NAME;
        playersFile = new File(playersPath);
        if (!playersFile.exists()) {
            findFile(playersFile, playersPath);
        }
    }

    /**
     * Reads a fixtures text file and stores its contents in a list
     */
    public static ArrayList<Fixture> readFixturesFile() {
        // Create a new BufferedReader that will go onto read the file's contents
        BufferedReader reader = findFile(fixturesFile, fixturesPath);

        // Stores the current line in the file
        String line;

        ArrayList<Fixture> fixtures = new ArrayList<>();

        try {
            // While the current line is not empty, read it and store it in the list
            while ((line = reader.readLine()) != null) {
                String homeTeamName = line;
                String awayTeamName = reader.readLine();
                int homeScore = Integer.parseInt(reader.readLine());
                int awayScore = Integer.parseInt(reader.readLine());
                ArrayList<String> goalscorers = new ArrayList<>();
                while (!((line = reader.readLine()).equals("-"))) {
                    goalscorers.add(line);
                }
                ArrayList<String> assists = new ArrayList<>();
                while (!((line = reader.readLine()).equals("-"))) {
                    assists.add(line);
                }
                int year = Integer.parseInt(reader.readLine());
                int month = Integer.parseInt(reader.readLine());
                int day = Integer.parseInt(reader.readLine());
                int hour = Integer.parseInt(reader.readLine());
                int minute = Integer.parseInt(reader.readLine());

                fixtures.add(new Fixture(homeTeamName, awayTeamName, new Score(homeScore, awayScore),
                        goalscorers, assists, LocalDateTime.of(year, month, day, hour, minute)));
            }
            reader.close();
        } catch (NullPointerException | IOException e) {
            Log.e(TAG, "Couldn't read fixtures file. The file may be empty");
        }
        Log.i(TAG, "Successfully read fixtures file");
        return fixtures;
    }


    /**
     * Reads a team text file and returns its contents
     */
    public static String readTeamFile() {
        // Create a new BufferedReader that will go onto read the file's contents
        BufferedReader reader = findFile(teamFile, teamPath);

        // Stores the current line in the file
        String line;

        String teamName = "";

        try {
            // While the current line is not empty, read it and store it in the list
            while ((line = reader.readLine()) != null) {
                teamName = line;
            }
            reader.close();
        } catch (NullPointerException | IOException e) {
            Log.e(TAG, "Couldn't read team file. The file may be empty");
        }
        Log.i(TAG, "Successfully read team file");
        return teamName;
    }


    /**
     * Reads a players text file and returns its contents
     */
    public static ArrayList<Player> readPlayersFile() {
        // Create a new BufferedReader that will go onto read the file's contents
        BufferedReader reader = findFile(playersFile, playersPath);

        // Stores the current line in the file
        String line;

        ArrayList<Player> players = new ArrayList<>();
        String playerName;

        try {
            // While the current line is not empty, read it and store it in the list
            while ((line = reader.readLine()) != null) {
                playerName = line;
                players.add(new Player(playerName));
            }
            reader.close();
        } catch (NullPointerException | IOException e) {
            Log.e(TAG, "Couldn't read players file. The file may be empty");
        }
        Log.i(TAG, "Successfully read team file");
        return players;
    }


    public static BufferedReader findFile(File file, String path) {
        BufferedReader reader = null;
        try {
            reader = new BufferedReader(new FileReader(path));
        } catch (FileNotFoundException e) {
            Log.i(TAG, "Couldn't find file. Creating the file...");
            try {
                // If the file does not exist, try and create the file
                file.getParentFile().mkdirs();
                file.createNewFile();
            } catch (IOException e2) {
                Log.e(TAG, "Couldn't create file");
            }
            Log.i(TAG, "Created file");
        }
        return reader;
    }


    /**
     * Write a team name to a file
     *
     * @param teamName The name of the team
     */
    public static void writeTeamFile(String teamName) {
        try {
            // Create a new BufferedWrite to write to the file
            BufferedWriter writer = new BufferedWriter(new FileWriter(new File(teamPath), false));
            // For every item in the list, write it to the file and then add a line break
            writer.write(teamName);
            writer.close();
        } catch (FileNotFoundException f) {
            /* If the file cannot be found try to read the file.
                This will include creating the file if it does not already exist,
                then attempt to write to the file again
             */
            readTeamFile();
            writeTeamFile(teamName);
        } catch (IOException e) {
            Log.e(TAG, e.getMessage());
        }
    }


    /**
     * Write a list of players to a file
     *
     * @param players The list of players
     */
    public static void writePlayersFile(ArrayList<Player> players) {
        try {
            // Create a new BufferedWrite to write to the file
            BufferedWriter writer = new BufferedWriter(new FileWriter(new File(playersPath), false));
            // For every player in the list, write it to the file and then add a line break
            for (Player player : players) {
                writer.write(player.getName());
                writer.newLine();
            }
            writer.close();
        } catch (FileNotFoundException f) {
            /* If the file cannot be found try to read the file.
                This will include creating the file if it does not already exist,
                then attempt to write to the file again
             */
            readPlayersFile();
            writePlayersFile(players);
        } catch (IOException e) {
            Log.e(TAG, e.getMessage());
        }
    }


    /**
     * Write a list of fixtures to a file
     *
     * @param fixtures The list of fixtures
     */
    public static void writeFixturesFile(ArrayList<Fixture> fixtures) {
        try {
            // Create a new BufferedWrite to write to the file
            BufferedWriter writer = new BufferedWriter(new FileWriter(new File(fixturesPath), false));
            // For every fixture in the list, write it to the file and then add a line break
            for (Fixture fixture : fixtures) {
                // Writes home team name
                writer.write(fixture.getHomeTeam());
                writer.newLine();
                // Writes away team name
                writer.write(fixture.getAwayTeam());
                writer.newLine();
                // Writes home score
                writer.write(Integer.toString(fixture.getScore().getHome()));
                writer.newLine();
                // Writes away score
                writer.write(Integer.toString(fixture.getScore().getAway()));
                writer.newLine();
                // Writes goalscorers
                ArrayList<String> goalscorers = fixture.getGoalscorers();
                for (String scorer : goalscorers) {
                    writer.write(scorer);
                    writer.newLine();
                }
                // Writes - to signify end of goalscorers
                writer.write("-");
                writer.newLine();
                // Writes assists
                ArrayList<String> assists = fixture.getAssists();
                for (String assist : assists) {
                    writer.write(assist);
                    writer.newLine();
                }
                // Writes - to signify end of assists
                writer.write("-");
                writer.newLine();
                // Writes year
                writer.write(Integer.toString(fixture.getDateTime().getYear()));
                writer.newLine();
                // Writes month
                writer.write(Integer.toString(fixture.getDateTime().getMonthValue()));
                writer.newLine();
                // Writes day
                writer.write(Integer.toString(fixture.getDateTime().getDayOfMonth()));
                writer.newLine();
                // Writes hour
                writer.write(Integer.toString(fixture.getDateTime().getHour()));
                writer.newLine();
                // Writes minute
                writer.write(Integer.toString(fixture.getDateTime().getMinute()));
                writer.newLine();
            }
            writer.close();
        } catch (FileNotFoundException f) {
            /* If the file cannot be found try to read the file.
                This will include creating the file if it does not already exist,
                then attempt to write to the file again
             */
            readFixturesFile();
            writeFixturesFile(fixtures);
        } catch (IOException e) {
            Log.e(TAG, e.getMessage());
        }
    }

}

File: app/src/main/java/com/jameschamberlain/footballteamtracker/FileUtils.kt
package com.jameschamberlain.footballteamtracker

import android.content.Context
import android.util.Log
import com.jameschamberlain.footballteamtracker.fixtures.Fixture
import com.jameschamberlain.footballteamtracker.fixtures.Score
import java.io.*
import java.time.LocalDateTime
import java.util.*

object FileUtils {
    private const val TAG = "FileUtils"
    private const val FIXTURES_FILE_NAME = "fixtures.txt"
    private const val TEAM_FILE_NAME = "team.txt"
    private const val PLAYERS_FILE_NAME = "players.txt"
    private var fixturesFile: File? = null
    private var teamFile: File? = null
    private var playersFile: File? = null
    private lateinit var fixturesPath: String
    private lateinit var teamPath: String
    private lateinit var playersPath: String


    fun checkFiles(context: Context) {
        fixturesPath = context.filesDir.toString() + FIXTURES_FILE_NAME
        fixturesFile = File(fixturesPath)
        if (fixturesFile!!.exists()) {
            findFile(fixturesFile!!, fixturesPath)
        }
        teamPath = context.filesDir.toString() + TEAM_FILE_NAME
        teamFile = File(teamPath)
        if (teamFile!!.exists()) {
            findFile(teamFile!!, teamPath)
        }
        playersPath = context.filesDir.toString() + PLAYERS_FILE_NAME
        playersFile = File(playersPath)
        if (playersFile!!.exists()) {
            findFile(playersFile!!, playersPath)
        }
    }

    /**
     * Reads a fixtures text file and stores its contents in a list
     */
    fun readFixturesFile(): ArrayList<Fixture> {
        // Create a new BufferedReader that will go onto read the file's contents
        val reader = findFile(fixturesFile!!, fixturesPath)

        // Stores the current line in the file
        var line: String?
        val fixtures = ArrayList<Fixture>()
        try {
            // While the current line is not empty, read it and store it in the list
            while (reader!!.readLine().also { line = it } != null) {
                val homeTeamName = line!!
                val awayTeamName = reader.readLine()
                val homeScore = reader.readLine().toInt()
                val awayScore = reader.readLine().toInt()
                val goalscorers = ArrayList<String>()
                while (reader.readLine().also { line = it } != "-") {
                    goalscorers.add(line!!)
                }
                val assists = ArrayList<String>()
                while (reader.readLine().also { line = it } != "-") {
                    assists.add(line!!)
                }
                val year = reader.readLine().toInt()
                val month = reader.readLine().toInt()
                val day = reader.readLine().toInt()
                val hour = reader.readLine().toInt()
                val minute = reader.readLine().toInt()
                fixtures.add(Fixture(homeTeamName, awayTeamName, Score(homeScore, awayScore),
                        goalscorers, assists, LocalDateTime.of(year, month, day, hour, minute)))
            }
            reader.close()
        } catch (e: NullPointerException) {
            Log.e(TAG, "Couldn't read fixtures file. The file may be empty")
        } catch (e: IOException) {
            Log.e(TAG, "Couldn't read fixtures file. The file may be empty")
        }
        Log.i(TAG, "Successfully read fixtures file")
        return fixtures
    }

    /**
     * Reads a team text file and returns its contents
     */
    fun readTeamFile(): String? {
        // Create a new BufferedReader that will go onto read the file's contents
        val reader = findFile(teamFile!!, teamPath)

        // Stores the current line in the file
        var line: String?
        var teamName: String? = ""
        try {
            // While the current line is not empty, read it and store it in the list
            while (reader!!.readLine().also { line = it } != null) {
                teamName = line
            }
            reader.close()
        } catch (e: NullPointerException) {
            Log.e(TAG, "Couldn't read team file. The file may be empty")
        } catch (e: IOException) {
            Log.e(TAG, "Couldn't read team file. The file may be empty")
        }
        Log.i(TAG, "Successfully read team file")
        return teamName
    }

    /**
     * Reads a players text file and returns its contents
     */
    fun readPlayersFile(): ArrayList<Player> {
        // Create a new BufferedReader that will go onto read the file's contents
        val reader = findFile(playersFile!!, playersPath)

        // Stores the current line in the file
        var line: String?
        val players = ArrayList<Player>()
        var playerName: String?
        try {
            // While the current line is not empty, read it and store it in the list
            while (reader!!.readLine().also { line = it } != null) {
                playerName = line
                players.add(Player(playerName!!))
            }
            reader.close()
        } catch (e: NullPointerException) {
            Log.e(TAG, "Couldn't read players file. The file may be empty")
        } catch (e: IOException) {
            Log.e(TAG, "Couldn't read players file. The file may be empty")
        }
        Log.i(TAG, "Successfully read team file")
        return players
    }

    private fun findFile(file: File?, path: String): BufferedReader? {
        var reader: BufferedReader? = null
        try {
            reader = BufferedReader(FileReader(path))
        } catch (e: FileNotFoundException) {
            Log.i(TAG, "Couldn't find file. Creating the file...")
            try {
                // If the file does not exist, try and create the file
                file?.parentFile?.mkdirs()
                file?.parentFile?.createNewFile()
            } catch (e2: IOException) {
                Log.e(TAG, "Couldn't create file")
            }
            Log.i(TAG, "Created file")
        }
        return reader
    }

    /**
     * Write a team name to a file
     *
     * @param teamName The name of the team
     */
    @JvmStatic
    fun writeTeamFile(teamName: String) {
        try {
            // Create a new BufferedWrite to write to the file
            val writer = BufferedWriter(FileWriter(File(teamPath), false))
            // For every item in the list, write it to the file and then add a line break
            writer.write(teamName)
            writer.close()
        } catch (f: FileNotFoundException) {
            /* If the file cannot be found try to read the file.
                This will include creating the file if it does not already exist,
                then attempt to write to the file again
             */
            readTeamFile()
            writeTeamFile(teamName)
        } catch (e: IOException) {
            Log.e(TAG, e.message!!)
        }
    }

    /**
     * Write a list of players to a file
     *
     * @param players The list of players
     */
    @JvmStatic
    fun writePlayersFile(players: ArrayList<Player>) {
        try {
            // Create a new BufferedWrite to write to the file
            val writer = BufferedWriter(FileWriter(File(playersPath), false))
            // For every player in the list, write it to the file and then add a line break
            for (player in players) {
                writer.write(player.name)
                writer.newLine()
            }
            writer.close()
        } catch (f: FileNotFoundException) {
            /* If the file cannot be found try to read the file.
                This will include creating the file if it does not already exist,
                then attempt to write to the file again
             */
            readPlayersFile()
            writePlayersFile(players)
        } catch (e: IOException) {
            Log.e(TAG, e.message!!)
        }
    }

    /**
     * Write a list of fixtures to a file
     *
     * @param fixtures The list of fixtures
     */
    @JvmStatic
    fun writeFixturesFile(fixtures: ArrayList<Fixture>) {
        try {
            // Create a new BufferedWrite to write to the file
            val writer = BufferedWriter(FileWriter(File(fixturesPath), false))
            // For every fixture in the list, write it to the file and then add a line break
            for (fixture in fixtures) {
                // Writes home team name
                writer.write(fixture.homeTeam)
                writer.newLine()
                // Writes away team name
                writer.write(fixture.awayTeam)
                writer.newLine()
                // Writes home score
                writer.write(fixture.score.home.toString())
                writer.newLine()
                // Writes away score
                writer.write(fixture.score.away.toString())
                writer.newLine()
                // Writes goalscorers
                val goalscorers = fixture.goalscorers
                for (scorer in goalscorers) {
                    writer.write(scorer)
                    writer.newLine()
                }
                // Writes - to signify end of goalscorers
                writer.write("-")
                writer.newLine()
                // Writes assists
                val assists = fixture.assists
                for (assist in assists) {
                    writer.write(assist)
                    writer.newLine()
                }
                // Writes - to signify end of assists
                writer.write("-")
                writer.newLine()
                // Writes year
                writer.write(fixture.dateTime.year.toString())
                writer.newLine()
                // Writes month
                writer.write(fixture.dateTime.monthValue.toString())
                writer.newLine()
                // Writes day
                writer.write(fixture.dateTime.dayOfMonth.toString())
                writer.newLine()
                // Writes hour
                writer.write(fixture.dateTime.hour.toString())
                writer.newLine()
                // Writes minute
                writer.write(fixture.dateTime.minute.toString())
                writer.newLine()
            }
            writer.close()
        } catch (f: FileNotFoundException) {
            /* If the file cannot be found try to read the file.
                This will include creating the file if it does not already exist,
                then attempt to write to the file again
             */
            readFixturesFile()
            writeFixturesFile(fixtures)
        } catch (e: IOException) {
            Log.e(TAG, e.message!!)
        }
    }
}